### ISSUE TITLE
Can't use Redis via Unix socket

### ISSUE THREAD

**Issue by gabrielelucci** [tootsuite/mastodon#5183]
Hi, I would like to make mastodon use Redis via Unix socket instead of TCP, but I can't seem to make it work. I found this closed issue https://github.com/tootsuite/mastodon/issues/1325, but, despite @Gargron claims that it's possible by setting REDIS_URL variable to some not-well-defined value, it just won't work.

What I tried:

- `REDIS_URL=redis:///var/run/redis/redis.sock`
- `REDIS_URL=redis://127.0.0.1/var/run/redis/redis.sock`
- `REDIS_URL=redis+unix:///var/run/redis/redis.sock`
- `REDIS_URL=redis+socket:///var/run/redis/redis.sock`

None of the above worked.

My setup:
Mastodon version: v1.6.1
Server: Ubuntu 16.04.3
Redis version: Redis server v=3.0.6

I also carefully examined the source code and the documentation, but didn't find anything useful. In fact, I think it's not implemented yet, doesn't work properly or it's buried somewhere and not documented.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by cooox** [comment]
TL;DR the correct scheme for REDIS_URL with a socket is unix://, use **REDIS_URL=unix:///var/run/redis/redis.sock**. Don't forget to enable the socket in the redis config and set the right permissions so the Mastodon user can read and write to the socket file.

How to find the error: if you enable Redis socket and disable the TCP socket on localhost, Mastodon will throw an Error 500 on the timeline. Inspecting the response you can see an error caused by `lib/redis/client.rb`. Digging deeper, I found that the environment variable REDIS_URL was correctly read, but had the wrong scheme (redis://). The solution is in [the client.rb](https://github.com/redis/redis-rb/blob/27165e22099a950bf4c3aac56684b42719284947/lib/redis/client.rb#L406), it shows you need the scheme unix:// for sockets. Or else host and port will be used, which is localhost TCP.

**Comment by gabrielelucci** [comment]
@cooox Nice! That's exactly the only scheme I didn't try! :P
I'll try that now, but I still think it should be better documented somewhere.

Closing this.

**Comment by gabrielelucci** [comment]
It worked! Don't know if it is the best way to do it, but I had to add the `mastodon` user to the `redis` group.

**Comment by ThisIsMissEm** [comment]
@gabrielelucci we've an upcoming change to the handling of `REDIS_URL` in 4.3 that's changing it to work inline with the way ruby works, would you be interested in testing this out to let us know if unix sockets for redis still works correctly for you?

**Comment by ThisIsMissEm** [comment]
Alrighty; I was just looking for anyone using the unix sockets for redis for this.

I actually theorize in #29066 that no one actually does use unix sockets for Streaming (neither for listen or redis or postgresql), but apparently people are downvoting it, without explaining their setups.

**Comment by Morethanevil** [comment]
I use unixsockets, but in a different scenario. I run mastodon in docker and I am passing the sockets file into the container. Instead of redis, I am running [Valkey](https://github.com/valkey-io/valkey) since it stays open source

Unixsockets are much faster than TCP connections.

**Comment by StandingPadAnimations** [comment]
> Alrighty; I was just looking for anyone using the unix sockets for redis for this.
> 
> I actually theorize in [#29066](https://github.com/mastodon/mastodon/issues/29066) that no one actually does use unix sockets for Streaming (neither for listen or redis or postgresql), but apparently people are downvoting it, without explaining their setups.

I know this is from two years ago, but I'd also like to explain my set-up: I use UNIX sockets because for rootless container set-ups, it's much, much slower to go through the TCP networking stack than through UNIX sockets.
